internal/graph/track_check_resolver_test.go

```diff
@@ -0,0 +1,280 @@
+			context.Background(),
+			model,
+		)
+		require.NoError(t, err)
+
+		ctx = storage.ContextWithRelationshipTupleReader(ctx, ds)
+		ctx = typesystem.ContextWithTypesystem(ctx, typesys)
+		resp, err := trackChecker.ResolveCheck(ctx, &ResolveCheckRequest{
+			AuthorizationModelID: ulid.Make().String(),
+			StoreID:              storeID,
+			TupleKey:             tuple.NewTupleKey("group:1", "blocked", "user:jon"),
+			RequestMetadata:      NewCheckRequestMetadata(25),
+		})
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("tracker_delegates_request", func(t *testing.T) {
+		ctrl := gomock.NewController(t)
+		t.Cleanup(ctrl.Finish)
+
+		mockLocalChecker := NewMockCheckResolver(ctrl)
+		mockLocalChecker.EXPECT().ResolveCheck(gomock.Any(), gomock.Any()).Return(&ResolveCheckResponse{
+			Allowed: true,
+		}, nil).Times(1)
+		trackChecker.SetDelegate(mockLocalChecker)
+
+		resp, err := trackChecker.ResolveCheck(ctx, &ResolveCheckRequest{
+			StoreID:         ulid.Make().String(),
+			TupleKey:        tuple.NewTupleKey("document:1", "viewer", "user:will"),
+			RequestMetadata: NewCheckRequestMetadata(defaultResolveNodeLimit),
+			VisitedPaths:    map[string]struct{}{},
+		})
+
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+
+	t.Run("tracker_user_type_validate_and_expiration", func(t *testing.T) {
+		require.True(t, trackChecker.validate())
+
+		tc := &TrackerCheckResolver{}
+		require.False(t, tc.validate())
+
+		tc = &TrackerCheckResolver{ctx: context.Background()}
+		require.False(t, tc.validate())
+
+		userType := trackChecker.userType("group:1#member")
+		require.Equal(t, "userset", userType)
+
+		userType = trackChecker.userType("user:ann")
+		require.Equal(t, "user", userType)
+
+		userType = trackChecker.userType("user:*")
+		require.Equal(t, "userset", userType)
+
+		r := resolutionNode{tm: time.Now().Add(-trackerInterval)}
+		require.True(t, r.expired())
+
+		r = resolutionNode{tm: time.Now().Add(trackerInterval)}
+		require.False(t, r.expired())
+	})
+
+	t.Run("tracker_prints_and_delete_path", func(t *testing.T) {
+		r := &ResolveCheckRequest{
+			StoreID:              ulid.Make().String(),
+			AuthorizationModelID: ulid.Make().String(),
+			TupleKey:             tuple.NewTupleKey("document:abc", "viewer", "user:somebody"),
+			RequestMetadata:      NewCheckRequestMetadata(20),
+		}
+		value, ok := trackChecker.loadModel(r)
+		require.NotNil(t, value)
+		require.False(t, ok)
+
+		path := "group:1#member@user"
+		sm := &sync.Map{}
+		sm.Store(
+			path,
+			&resolutionNode{
+				tm:   time.Now().Add(-trackerInterval),
+				hits: &atomic.Uint64{},
+			},
+		)
+
+		trackChecker.nodes.Store(trackerKey{store: ulid.Make().String(), model: ulid.Make().String()}, sm)
+		trackChecker.logExecutionPaths(false)
+
+		_, ok = sm.Load(path)
+		require.False(t, ok)
+	})
+
+	t.Run("tracker_success_launch_flush", func(t *testing.T) {
```

Comment:
I have too many questions about this test 😅 Starting with the name. If you called `launchFlush` and didn't get any errors back, that should be enough. But since `launchFlush` doesn't return an error, then you need to observe its effects, right? So `launchFlush` writes to a log. But that only happens at certain time intervals. So why do you create a goroutine that sleeps for 10 ms and then wait on it? This smells like a test that will be flaky. Why not make the time interval zero (maybe an optional parameter to the constructor) so that as soon as there is an event, it flushes it, and you don't have to wait on it?